datasource/opennms-kafka/src/test/java/org/opennms/alec/datasource/opennms/KafkaRecoveryTest.java

```diff
@@ -0,0 +1,151 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.alec.datasource.opennms;
+
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.hasSize;
+
+import java.lang.reflect.Method;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.streams.processor.internals.StreamThread;
+import org.junit.Test;
+import org.opennms.alec.datasource.api.Alarm;
+import org.opennms.alec.datasource.api.AlarmHandler;
+import org.opennms.alec.datasource.opennms.proto.OpennmsModelProtos;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import kafka.server.KafkaServer;
+
+/** Makes sure that we can recover after the connection to Kafka is lost. */
+public class KafkaRecoveryTest extends OpennmsDatasourceIT implements AlarmHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaRecoveryTest.class);
+
+    private final Map<String, Alarm> alarmsCreatedOrUpdatedById = new LinkedHashMap<>();
+    private final Map<String, Alarm> alarmsClearedById = new LinkedHashMap<>();
+
+    @Test(timeout=60000)
+    public void canHandleAlarms() throws Exception {
+        datasource.init();
+
+        // No alarms initially
+        assertThat(datasource.getAlarms(), hasSize(0));
+
+        // Send one alarm
+        OpennmsModelProtos.Alarm alarm = OpennmsModelProtos.Alarm.newBuilder()
+                .setReductionKey("nodeDown::1")
+                .setLastEventTime(1)
+                .setSeverity(OpennmsModelProtos.Severity.CRITICAL)
+                .build();
+        producer.send(new ProducerRecord<>(datasource.getAlarmTopic(), alarm.getReductionKey(), alarm.toByteArray()));
+
+        // Expect a single alarm to be present in the datasource
+        await().atMost(10, TimeUnit.SECONDS).until(() -> datasource.getAlarms(), hasSize(1));
+        // Now register a handler
+        assertThat(datasource.getAlarmsAndRegisterHandler(this), hasSize(1));
+        // We should not have received any callbacks yet
+        assertThat(alarmsCreatedOrUpdatedById.entrySet(), hasSize(0));
+        assertThat(alarmsClearedById.entrySet(), hasSize(0));
+
+        // Simulate a problem with KafkaStreams in OpennmsDatsource:
+        // Stop all StreamsThreads and wait for the KafkaStreamMonitor to restart the OpennmsDatasource
+        LOG.info("Trigger KafkaStream problem.");
+        // "Kill" all threads
+        for (Thread t : Thread.getAllStackTraces().keySet()) {
```

Comment:
The test doesn't see to work as expected. It try here to kill all StreamThreads which belong to  OpennmsDatasource.streams. The killing seems to work all right but it looks like we can't recover from that by calling OpennmsDatasource.init() which is what the KafkaStreamMonitor is doing. I assume the failure is due to the test setup but I am not sure how to fix it.

@j-white we should discuss 
* if this is the right approach and 
* how to fix the test.